### PR TITLE
Add reverse infer

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *Add transformation for deducing rank of input by reversing shape inference*
 - *Support Convolution reverse infer*
 - *Support Pad reverse infer*

### Tickets:
 - *98182*

### CODE REVIEW

File: src/common/transformations/include/transformations/common_optimizations/reverse_shape_and_type_infer.hpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#include <vector>
+
+namespace ov {
+namespace pass {
+
+class TRANSFORMATIONS_API ReverseShapeAndTypeInfer;
+
+}  // namespace pass
+}  // namespace ov
+
```
Comment (rkazants) [inline src/common/transformations/include/transformations/common_optimizations/reverse_shape_and_type_infer.hpp:19]: Please add description for the transformation

File: src/common/transformations/include/transformations/common_optimizations/reverse_shape_and_type_infer.hpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+class ov::pass::ReverseShapeAndTypeInfer : public ngraph::pass::FunctionPass {
+public:
+    OPENVINO_RTTI("ReverseShapeAndTypeInfer", "0");
+    bool run_on_model(const std::shared_ptr<ngraph::Function>& f) override;
```
Comment (rkazants) [inline src/common/transformations/include/transformations/common_optimizations/reverse_shape_and_type_infer.hpp:23]: please use `ov::Model`. Please avoid `ngraph` use in your code

File: src/core/src/op/pad.cpp [hunk 3]
```diff
@@ -102,7 +110,7 @@ void op::v1::Pad::validate_and_infer_types() {
     for (size_t i = 0; i < get_input_size(); i++)
         input_shapes.push_back(get_input_partial_shape(i));
     shape_infer(this, input_shapes, output_shapes);
-    set_output_type(0, get_input_element_type(0), output_shapes[0]);
+    set_output_type(0, result_et, output_shapes[0]);
```
Comment (rkazants) [inline src/core/src/op/pad.cpp:113]: strange change in the core. I think we should not expect such changes in the core part

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+namespace {
+bool inherit_output_shape(std::shared_ptr<ov::Node> node, std::vector<size_t> input_idxs) {
+    auto is_changed = false;
+    auto output_shape = node->get_output_partial_shape(0);
+
+    for (auto idx : input_idxs) {
+        if (node->get_input_partial_shape(idx).rank().is_dynamic()) {
+            OPENVINO_SUPPRESS_DEPRECATED_START
+            node->get_input_tensor(idx).set_partial_shape(output_shape);
+            OPENVINO_SUPPRESS_DEPRECATED_END
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:28]: @ilyachur, please pay attention to that we use publicly deprecated methods for reverse infer functionality. It is required for our purposes.

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            is_changed = true;
+        }
+    }
+    return is_changed;
+}
+
+bool inherit_output_rank(std::shared_ptr<ov::Node> node, std::vector<size_t> input_idxs) {
+    auto is_changed = false;
+    auto output_shape = node->get_output_partial_shape(0);
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:37]: may be it is better to pass the certain rank size to this method and set it up? This way we can use this routine in more cases not only when input rank size equal to output rank size.

File: src/core/src/op/pad.cpp (hunk 3 above)
Comment (mvafin) [reply src/core/src/op/pad.cpp:111]: That was found by tests and seems logical. For `Pad` in constant mode we may have value and the type of value should be the same as data type, so we can set such type as type of `Pad` output. And if we do reverse infer we can set this type to `Pad` input as well.

File: src/core/src/op/pad.cpp [hunk 6]
```diff
@@ -66,12 +66,20 @@ bool ngraph::op::v1::Pad::visit_attributes(AttributeVisitor& visitor) {
...
     const auto& pads_end_element_type = get_input_element_type(2);
 
+    NODE_VALIDATION_CHECK(this,
+                          element::Type::merge(result_et, result_et, arg_element_type),
+                          "Cannot merge element types (input arg element type: ",
+                          arg_element_type,
+                          ", with: ",
+                          result_et,
+                          ").");
+
```
Comment (rkazants) [inline src/core/src/op/pad.cpp:82]: please add a test for Pad shape_inference to cover this case

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp (hunk 4 above)
Comment (rkazants) [reply src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:21]: @ilyachur, let us consider possibility to un-deprecate `set_partial_shape()` method

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+bool ov::pass::ReverseShapeAndTypeInfer::run_on_model(const std::shared_ptr<ov::Model>& f) {
+    RUN_ON_MODEL_SCOPE(ReverseShapeAndTypeInfer);
+    bool is_changed = false;
+    auto ops = f->get_ordered_ops();
+    OPENVINO_SUPPRESS_DEPRECATED_START
+    for (auto it = ops.rbegin(); it != ops.rend(); ++it) {
+        const auto& op = *it;
+        auto output_shape = op->get_output_partial_shape(0);
+        auto output_type = op->get_output_element_type(0);
+        std::cout << op->get_type_name() << std::endl;
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:68]: please remove `std::cout`

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        if (const auto& param = std::dynamic_pointer_cast<Parameter>(op)) {
+            if (param->get_partial_shape().rank().is_dynamic()) {
+                param->set_partial_shape(output_shape);
+                is_changed = true;
+            }
+            if (param->get_element_type().is_dynamic()) {
+                param->set_element_type(output_type);
+                is_changed = true;
+            }
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:77]: looks not needed step. `output_shape` and `output_type` must be set by the next operation in the graph

**Review by rkazants — CHANGES_REQUESTED** [review]

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+#include "transformations/common_optimizations/reverse_shape_and_type_infer.hpp"
+
+#include "itt.hpp"
+#include "openvino/opsets/opset10.hpp"
+
+using namespace ov::opset10;
+
+namespace {
+bool inherit_output_shape(std::shared_ptr<ov::Node> node, std::vector<size_t> input_idxs) {
```
Comment (ilyachur) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:13]: Please avoid additional copies:
```suggestion
bool inherit_output_shape(const std::shared_ptr<ov::Node>& node, const std::vector<size_t>& input_idxs) {
```

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+bool inherit_output_rank(std::shared_ptr<ov::Node> node, std::vector<size_t> input_idxs) {
+    auto is_changed = false;
+    auto output_shape = node->get_output_partial_shape(0);
+
+    for (auto idx : input_idxs) {
+        if (idx < node->get_input_size() && node->get_input_partial_shape(idx).rank().is_dynamic()) {
+            OPENVINO_SUPPRESS_DEPRECATED_START
+            node->get_input_tensor(idx).set_partial_shape(ov::PartialShape::dynamic(output_shape.rank()));
```
Comment (ilyachur) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:35]: The same comment

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+bool inherit_output_type(std::shared_ptr<ov::Node> node, std::vector<size_t> input_idxs) {
+    auto is_changed = false;
+    auto output_type = node->get_output_element_type(0);
+
+    for (auto idx : input_idxs) {
+        if (node->get_input_element_type(idx).is_dynamic()) {
+            OPENVINO_SUPPRESS_DEPRECATED_START
+            node->get_input_tensor(idx).set_element_type(output_type);
```
Comment (ilyachur) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:50]: Please add convert or change element type of parameter op

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+namespace {
+bool inherit_output_shape(std::shared_ptr<ov::Node> node, std::vector<size_t> input_idxs) {
+    auto is_changed = false;
+    auto output_shape = node->get_output_partial_shape(0);
+
+    for (auto idx : input_idxs) {
+        if (node->get_input_partial_shape(idx).rank().is_dynamic()) {
+            OPENVINO_SUPPRESS_DEPRECATED_START
+            node->get_input_tensor(idx).set_partial_shape(output_shape);
```
[the remark this answers is not in the document]
Comment (mvafin) [reply src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:20]: That is exactly the purpose of this PR. It deduce rank of the input based on the output and op type. For example for Convolution we can deduce input rank when we know that weights are 4 dimensional. That is needed for the cases when we do not have information about input.

File: src/common/transformations/tests/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 13]
```diff
@@ -0,0 +1,199 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <ngraph/pass/manager.hpp>
```
Comment (rkazants) [inline src/common/transformations/tests/common_optimizations/reverse_shape_and_type_infer.cpp:8]: please avoid usage of ngraph routines

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp (hunk 11 above)
Comment (mvafin) [reply src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:50]: To do that we need to propagate the type and shape up. That will eventually happen on lines 69-77

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp (hunk 8 above)
Comment (mvafin) [reply src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:77]: This step is required. It sets type and shape to Parameter node. Without it `validate_and_infer_types` on whole model will remove all deduced ranks.

File: src/common/transformations/include/transformations/common_optimizations/reverse_shape_and_type_infer.hpp [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+namespace ov {
+namespace pass {
+
+class TRANSFORMATIONS_API ReverseShapeAndTypeInfer;
+
+}  // namespace pass
+}  // namespace ov
+
+/**
+ * @brief Perform reverse shape and type infer to duduce input rank and type in certain cases
```
Comment (rkazants) [inline src/common/transformations/include/transformations/common_optimizations/reverse_shape_and_type_infer.hpp:19]:
```suggestion
 * @brief Perform reverse shape and type infer to deduce input rank and type in certain cases
```

**Review by rkazants — APPROVED** [review]

**Pull request merged**